File: cpu/src/io.rs
use core::arch::asm;

/// Writes a byte to an I/O port
///
/// # Safety
/// Modifying I/O ports can have unintended side effects, and is inherently unsafe.
pub unsafe fn outb(port: u16, val: u8) {
    unsafe {
        asm!("out dx, al", in("dx") port, in("al") val);
    }
}

/// Writes 2 bytes to an I/O port
///
/// # Safety
/// Modifying I/O ports can have unintended side effects, and is inherently unsafe.
pub unsafe fn outw(port: u16, val: u16) {
    unsafe {
        asm!("out dx, ax", in("dx") port, in("ax") val);
    }
}

/// Writes 4 bytes to an I/O port
///
/// # Safety
/// Modifying I/O ports can have unintended side effects, and is inherently unsafe.
pub unsafe fn outl(port: u16, val: u32) {
    unsafe {
        asm!("out dx, eax", in("dx") port, in("eax") val);
    }
}

/// Reads a byte from an I/O port
///
/// # Safety
/// Reading from I/O ports can have unintended side effects, and is inherently unsafe.
pub unsafe fn inb(port: u16) -> u8 {
    let ret: u8;
    unsafe {
        asm!("in al, dx", out("al") ret, in("dx") port);
    }
    ret
}

/// Reads 2 bytes from an I/O port
///
/// # Safety
/// Reading from I/O ports can have unintended side effects, and is inherently unsafe.
pub unsafe fn inw(port: u16) -> u16 {
    let ret: u16;
    unsafe {
        asm!("in ax, dx", out("ax") ret, in("dx") port);
    }
    ret
}

/// Reads 4 bytes from an I/O port
///
/// # Safety
/// Reading from I/O ports can have unintended side effects, and is inherently unsafe.
pub unsafe fn inl(port: u16) -> u32 {
    let ret: u32;
    unsafe {
        asm!("in eax, dx", out("eax") ret, in("dx") port);
    }
    ret
}

File: kernel/src/io.c
#include "io.h"

inline void outb(uint16_t port, uint8_t val) {
    __asm__ volatile("outb %b0, %w1" : : "a"(val), "Nd"(port) : "memory");
}

inline void outw(uint16_t port, uint16_t val) {
    __asm__ volatile("outw %w0, %w1" : : "a"(val), "Nd"(port) : "memory");
}

inline void outl(uint16_t port, uint32_t val) {
    __asm__ volatile("outl %0, %w1" : : "a"(val), "Nd"(port) : "memory");
}

inline uint8_t inb(uint16_t port) {
    uint8_t ret;
    __asm__ volatile("inb %w1, %b0" : "=a"(ret) : "Nd"(port) : "memory");
    return ret;
}

inline uint16_t inw(uint16_t port) {
    uint16_t ret;
    __asm__ volatile("inw %w1, %w0" : "=a"(ret) : "Nd"(port) : "memory");
    return ret;
}

inline uint32_t inl(uint16_t port) {
    uint32_t ret;
    __asm__ volatile("inl %w1, %0" : "=a"(ret) : "Nd"(port) : "memory");
    return ret;
}

inline void io_wait() { outb(0x80, 0); }
